gordon_janitor_gcp/gpubsub_client.py

```diff
@@ -0,0 +1,224 @@
+        return auth.GoogleAuthClient(keyfile=keyfile, scopes=scopes)
+
+    def _init_publisher(self, auth_client):
+        try:
+            project = self.config['project']
+        except KeyError:
+            msg = 'The GCP project where Cloud Pub/Sub is located is required.'
+            logging.error(msg)
+            raise exceptions.GCPConfigError(msg)
+        try:
+            topic = self.config['topic']
+        except KeyError:
+            msg = ('A topic for the client to publish to in Cloud Pub/Sub is '
+                   'required.')
+            logging.error(msg)
+            raise exceptions.GCPConfigError(msg)
+
+        self.topic = f'projects/{project}/topics/{topic}'
+        return self._get_publisher(auth_client)
+
+    def _get_publisher(self, auth_client):
+        # Silly emulator constraints
+        creds = getattr(auth_client, 'creds', None)
+        client = pubsub.PublisherClient(credentials=creds)
+
+        try:
+            client.create_topic(self.topic)
+        except google_exceptions.AlreadyExists:
+            # already created
+            pass
+        except Exception as e:
+            msg = f'Error trying to create topic "{self.topic}": {e}'
+            logging.error(msg, exc_info=e)
+            raise exceptions.GCPGordonJanitorError(msg)
+        return client
+
+    async def done(self):
+        """Clean up outstanding tasks and close HTTP session.
+
+        This method collects all tasks that this particular class
+        initiated, and will cancel them if they don't complete within
+        the configured timeout period.
+        """
+        tasks_to_clear = [
+            t for t in self._messages if not t.done()
+        ]
+        sleep_for = 0.5  # half a second
+        iterations = self.cleanup_timeout / sleep_for
+
+        while iterations:
+            tasks_to_clear = [t for t in tasks_to_clear if not t.done()]
+            if not tasks_to_clear:
+                break
+
+            await asyncio.sleep(sleep_for)
+            iterations -= 1
+
+        # give up on waiting for tasks to complete
+        if tasks_to_clear:
+            msg = (f'The following tasks did not complete in time and are '
+                   f'being cancelled: {tasks_to_clear}')
+            logging.warning(msg)
+            for task in tasks_to_clear:
+                task.cancel()
+
+        # TODO (lynn): add metrics.flush call here once aioshumway is released
+        msg = ('Finished sending reconciliation messages to Google Pub/Sub.')
+        logging.info(msg)
+
+    @threads.threadpool
+    def publish(self, message):
+        """Publish received change message to Google Pub/Sub.
+
+        Args:
+            message (dict): change message received from the
+                ``self.changes_channel`` to emit.
+        """
+        message['timestamp'] = datetime.datetime.utcnow().isoformat()
+        bytes_message = bytes(json.dumps(message).encode('utf-8'))
```

Comment:
The call to `bytes()` here is a noop because `.encode('utf-8')` returns a bytes objects:

```python
>>> msgstr = json.dumps({'data': 123})
>>> bytes(msgstr, encoding='utf-8') == msgstr.encode('utf-8')
>>> True
```

Reply:
:derp: